#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
é¢ç›¸æ‰‹ç›¸åˆ†æ API æ¥å£ - æµå¼ç‰ˆæœ¬
æä¾›æµå¼ HTTP æ¥å£ï¼Œä½¿ç”¨ Server-Sent Events (SSE)
"""

import sys
import os
import json
import asyncio
import logging
from fastapi import APIRouter, UploadFile, File, Form, HTTPException, Header, Depends
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Optional
import grpc

logger = logging.getLogger(__name__)

# æ·»åŠ é¡¹ç›®æ ¹ç›®å½•åˆ°è·¯å¾„
# server/api/v1/fortune_analysis_stream.py -> å‘ä¸Š4çº§åˆ°é¡¹ç›®æ ¹
current_file = os.path.abspath(__file__)
project_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(current_file))))
sys.path.insert(0, project_root)

# å¯¼å…¥ç”Ÿæˆçš„ gRPC ä»£ç 
generated_path = os.path.join(project_root, "proto", "generated")
if generated_path not in sys.path:
    sys.path.insert(0, generated_path)

# ä½¿ç”¨ importlib ç›´æ¥ä»æ–‡ä»¶å¯¼å…¥ï¼Œé¿å…æ¨¡å—è·¯å¾„é—®é¢˜
import importlib.util

pb2_path = os.path.join(generated_path, "fortune_analysis_pb2.py")
pb2_grpc_path = os.path.join(generated_path, "fortune_analysis_pb2_grpc.py")

if os.path.exists(pb2_path) and os.path.exists(pb2_grpc_path):
    # ä»æ–‡ä»¶ç›´æ¥åŠ è½½æ¨¡å—
    spec = importlib.util.spec_from_file_location("fortune_analysis_pb2", pb2_path)
    fortune_analysis_pb2 = importlib.util.module_from_spec(spec)
    sys.modules["fortune_analysis_pb2"] = fortune_analysis_pb2
    spec.loader.exec_module(fortune_analysis_pb2)
    
    spec = importlib.util.spec_from_file_location("fortune_analysis_pb2_grpc", pb2_grpc_path)
    fortune_analysis_pb2_grpc = importlib.util.module_from_spec(spec)
    sys.modules["fortune_analysis_pb2_grpc"] = fortune_analysis_pb2_grpc
    spec.loader.exec_module(fortune_analysis_pb2_grpc)
else:
    # å¦‚æœæ–‡ä»¶ä¸å­˜åœ¨ï¼Œå°è¯•æ™®é€šå¯¼å…¥
    try:
        import fortune_analysis_pb2
        import fortune_analysis_pb2_grpc
    except ImportError as e:
        raise ImportError(f"æ— æ³•å¯¼å…¥ fortune_analysis_pb2ï¼Œæ–‡ä»¶è·¯å¾„: {pb2_path}, é”™è¯¯: {e}")

router = APIRouter()

# è·å–æœåŠ¡åœ°å€
FORTUNE_ANALYSIS_SERVICE_URL = os.getenv("FORTUNE_ANALYSIS_SERVICE_URL", "127.0.0.1:9005")

# ç®€å•è®¤è¯ï¼šAPI Keyï¼ˆå¿…é¡»ä»ç¯å¢ƒå˜é‡è¯»å–ï¼‰
# å®‰å…¨è§„èŒƒï¼šç”Ÿäº§ç¯å¢ƒä¸å…è®¸ä½¿ç”¨é»˜è®¤å€¼
_app_env = os.getenv("APP_ENV", "development").lower()
if _app_env == "production":
    FORTUNE_ANALYSIS_API_KEY = os.getenv("FORTUNE_ANALYSIS_API_KEY")
    if not FORTUNE_ANALYSIS_API_KEY:
        import logging
        logging.getLogger(__name__).warning(
            "âš ï¸ ç”Ÿäº§ç¯å¢ƒæœªé…ç½® FORTUNE_ANALYSIS_API_KEYï¼Œfortune_analysis æ¥å£å°†æ‹’ç»æ‰€æœ‰è¯·æ±‚"
        )
else:
    # å¼€å‘ç¯å¢ƒï¼šå…è®¸ä½¿ç”¨é»˜è®¤å€¼ï¼ˆæ–¹ä¾¿æœ¬åœ°è°ƒè¯•ï¼‰
    FORTUNE_ANALYSIS_API_KEY = os.getenv("FORTUNE_ANALYSIS_API_KEY", "dev_fortune_analysis_key")

# å¯¼å…¥å›¾åƒéªŒè¯å™¨å’Œæµå¼åˆ†æå™¨
# æ·»åŠ æœåŠ¡ç›®å½•åˆ°è·¯å¾„ï¼Œç¡®ä¿èƒ½æ­£ç¡®å¯¼å…¥
fortune_service_dir = os.path.join(project_root, "services", "fortune_analysis")
if fortune_service_dir not in sys.path:
    sys.path.insert(0, fortune_service_dir)

try:
    from image_validator import ImageValidator
    image_validator = ImageValidator()
    IMAGE_VALIDATOR_AVAILABLE = True
    logger.info("âœ… å›¾åƒéªŒè¯å™¨å¯¼å…¥æˆåŠŸ")
except ImportError as e:
    logger.warning(f"âš ï¸  å›¾åƒéªŒè¯å™¨å¯¼å…¥å¤±è´¥: {e}ï¼Œå°†è·³è¿‡å›¾åƒç±»å‹éªŒè¯")
    image_validator = None
    IMAGE_VALIDATOR_AVAILABLE = False

# å¯¼å…¥æµå¼åˆ†æå™¨ï¼ˆåˆ†åˆ«å¯¼å…¥æ‰‹ç›¸å’Œé¢ç›¸ï¼Œäº’ä¸å½±å“ï¼‰
try:
    from hand_analyzer_stream import HandAnalyzerStream
    hand_analyzer_stream = HandAnalyzerStream()
    HAND_STREAM_ANALYZER_AVAILABLE = True
    logger.info("âœ… æ‰‹ç›¸æµå¼åˆ†æå™¨å¯¼å…¥æˆåŠŸ")
except ImportError as e:
    import traceback
    logger.warning(f"âš ï¸  æ‰‹ç›¸æµå¼åˆ†æå™¨å¯¼å…¥å¤±è´¥: {e}", exc_info=True)
    hand_analyzer_stream = None
    HAND_STREAM_ANALYZER_AVAILABLE = False

try:
    from face_analyzer_stream import FaceAnalyzerStream
    face_analyzer_stream = FaceAnalyzerStream()
    FACE_STREAM_ANALYZER_AVAILABLE = True
    logger.info("âœ… é¢ç›¸æµå¼åˆ†æå™¨å¯¼å…¥æˆåŠŸ")
except ImportError as e:
    import traceback
    logger.warning(f"âš ï¸  é¢ç›¸æµå¼åˆ†æå™¨å¯¼å…¥å¤±è´¥: {e}", exc_info=True)
    face_analyzer_stream = None
    FACE_STREAM_ANALYZER_AVAILABLE = False


def get_grpc_client():
    """è·å– gRPC å®¢æˆ·ç«¯åœ°å€"""
    address = FORTUNE_ANALYSIS_SERVICE_URL
    if address.startswith("http://"):
        address = address[7:]
    elif address.startswith("https://"):
        address = address[8:]
    
    if ":" not in address:
        address = f"{address}:9005"
    
    return address


async def verify_api_key(x_api_key: Optional[str] = Header(None, alias="X-API-Key")):
    """
    éªŒè¯ API Keyï¼ˆç®€å•è®¤è¯ï¼‰
    
    Args:
        x_api_key: ä»è¯·æ±‚å¤´è·å–çš„ API Key
        
    Returns:
        API Key å­—ç¬¦ä¸²
        
    Raises:
        HTTPException: å¦‚æœ API Key æ— æ•ˆ
    """
    if not x_api_key or x_api_key != FORTUNE_ANALYSIS_API_KEY:
        raise HTTPException(
            status_code=401,
            detail="æ— æ•ˆçš„ API Keyï¼Œè¯·åœ¨è¯·æ±‚å¤´ä¸­æä¾›æ­£ç¡®çš„ X-API-Key"
        )
    return x_api_key


async def analyze_hand_stream_generator(image_bytes: bytes, image_format: str, 
                                        solar_date: Optional[str], solar_time: Optional[str], 
                                        gender: Optional[str], use_bazi: bool):
    """æ‰‹ç›¸åˆ†ææµå¼ç”Ÿæˆå™¨ï¼ˆçœŸæ­£çš„æµå¼å“åº”ï¼‰"""
    try:
        # å‘é€å¼€å§‹æ¶ˆæ¯ï¼ˆåŒ…å«æ˜¾ç¤ºæŒ‡ä»¤ï¼‰
        start_msg = {
            'type': 'start',
            'statusText': 'å¼€å§‹åˆ†ææ‰‹ç›¸',
            'showGrid': True,
            'showStatus': True,
            'showPlayButton': True
        }
        yield f"data: {json.dumps(start_msg, ensure_ascii=False)}\n\n"
        await asyncio.sleep(0)
        
        # ä½¿ç”¨æ‰‹ç›¸æµå¼åˆ†æå™¨ï¼ˆçœŸæ­£çš„æµå¼å“åº”ï¼‰
        if HAND_STREAM_ANALYZER_AVAILABLE and hand_analyzer_stream:
            logger.info("âœ… ä½¿ç”¨æ‰‹ç›¸æµå¼åˆ†æå™¨ï¼ˆçœŸæ­£çš„æµå¼å“åº”ï¼‰")
            try:
                # ä½¿ç”¨æ‰‹ç›¸æµå¼åˆ†æå™¨
                async for result in hand_analyzer_stream.analyze_hand_stream(
                    image_bytes=image_bytes,
                    image_format=image_format,
                    solar_date=solar_date,
                    solar_time=solar_time,
                    gender=gender,
                    use_bazi=use_bazi,
                    progress_callback=lambda text: None  # è¿›åº¦åœ¨å†…éƒ¨å¤„ç†
                ):
                    logger.debug(f"ğŸ“¤ æµå¼åˆ†æå™¨è¿”å›: type={result.get('type')}, statusText={result.get('statusText', '')}")
                    
                    # è½¬æ¢ç»“æœæ ¼å¼
                    if result.get('type') == 'progress':
                        progress_msg = {
                            'type': 'progress',
                            'statusText': result.get('statusText', 'æ­£åœ¨åˆ†æ...'),
                            'showGrid': True,
                            'showStatus': True,
                            'showPlayButton': True
                        }
                        yield f"data: {json.dumps(progress_msg, ensure_ascii=False)}\n\n"
                        await asyncio.sleep(0)
                    elif result.get('type') == 'complete':
                        data = result.get('data', {})
                        logger.debug(f"ğŸ“¤ å‘é€completeæ¶ˆæ¯ï¼Œdataç±»å‹: {type(data)}, success: {data.get('success') if isinstance(data, dict) else 'N/A'}")
                        complete_msg = {
                            'type': 'complete',
                            'data': data,
                            'statusText': 'åˆ†æå®Œæˆ',
                            'showGrid': False,
                            'showStatus': False,
                            'showPlayButton': False
                        }
                        yield f"data: {json.dumps(complete_msg, ensure_ascii=False)}\n\n"
                        logger.debug("âœ… completeæ¶ˆæ¯å·²å‘é€åˆ°å®¢æˆ·ç«¯")
                    elif result.get('type') == 'error':
                        error_msg = {
                            'type': 'error',
                            'error': result.get('error', 'åˆ†æå¤±è´¥'),
                            'statusText': 'åˆ†æå¤±è´¥',
                            'showGrid': False,
                            'showStatus': False,
                            'showPlayButton': False
                        }
                        yield f"data: {json.dumps(error_msg, ensure_ascii=False)}\n\n"
                        logger.error("âŒ erroræ¶ˆæ¯å·²å‘é€åˆ°å®¢æˆ·ç«¯")
            except Exception as stream_error:
                import traceback
                error_detail = f"æµå¼åˆ†æå™¨å¼‚å¸¸: {str(stream_error)}\n{traceback.format_exc()}"
                logger.error(f"âŒ {error_detail}")
                error_msg = {
                    'type': 'error',
                    'error': f"æµå¼åˆ†æå¤±è´¥: {str(stream_error)}",
                    'statusText': 'åˆ†æå¤±è´¥',
                    'showGrid': False,
                    'showStatus': False,
                    'showPlayButton': False
                }
                yield f"data: {json.dumps(error_msg, ensure_ascii=False)}\n\n"
        else:
            logger.warning("âš ï¸  æµå¼åˆ†æå™¨ä¸å¯ç”¨ï¼Œä½¿ç”¨ gRPCï¼ˆéæµå¼ï¼Œè¾ƒæ…¢ï¼‰")
            # é™çº§æ–¹æ¡ˆï¼šä½¿ç”¨ gRPCï¼ˆéæµå¼ï¼‰
            # å…ˆéªŒè¯å›¾åƒ
            if IMAGE_VALIDATOR_AVAILABLE and image_validator:
                logger.debug("ğŸ” éªŒè¯å›¾åƒ...")
                is_valid, error_msg = image_validator.validate_hand_image(image_bytes)
                logger.debug(f"éªŒè¯ç»“æœ: æœ‰æ•ˆ={is_valid}, é”™è¯¯={error_msg}")
                if not is_valid:
                    error_response = {
                        'type': 'error',
                        'error': error_msg,
                        'statusText': 'å›¾åƒéªŒè¯å¤±è´¥',
                        'showGrid': False,
                        'showStatus': False,
                        'showPlayButton': False
                    }
                    yield f"data: {json.dumps(error_response, ensure_ascii=False)}\n\n"
                    return
            else:
                logger.warning("âš ï¸  å›¾åƒéªŒè¯å™¨ä¸å¯ç”¨ï¼Œè·³è¿‡éªŒè¯")
            
            # æ„å»º gRPC è¯·æ±‚
            request = fortune_analysis_pb2.HandAnalysisRequest()
            request.image.image_bytes = image_bytes
            request.image.image_format = image_format
            request.image.filename = f"hand.{image_format}"
            
            # å…«å­—ä¿¡æ¯
            if use_bazi and solar_date and solar_time and gender:
                request.bazi_info.solar_date = solar_date
                request.bazi_info.solar_time = solar_time
                request.bazi_info.gender = gender
                request.bazi_info.use_bazi = True
            else:
                request.bazi_info.use_bazi = False
            
            # è°ƒç”¨ gRPC æœåŠ¡
            address = get_grpc_client()
            
            with grpc.insecure_channel(address) as channel:
                stub = fortune_analysis_pb2_grpc.FortuneAnalysisServiceStub(channel)
                response = stub.AnalyzeHand(request, timeout=60.0)
            
            if response.success:
                report = json.loads(response.report_json) if response.report_json else {}
                complete_msg = {
                    'type': 'complete',
                    'data': report,
                    'statusText': 'åˆ†æå®Œæˆ',
                    'showGrid': False,
                    'showStatus': False,
                    'showPlayButton': False
                }
                yield f"data: {json.dumps(complete_msg, ensure_ascii=False)}\n\n"
            else:
                error_msg = {
                    'type': 'error',
                    'error': response.error_message,
                    'statusText': 'åˆ†æå¤±è´¥',
                    'showGrid': False,
                    'showStatus': False,
                    'showPlayButton': False
                }
                yield f"data: {json.dumps(error_msg, ensure_ascii=False)}\n\n"
            
    except Exception as e:
        import traceback
        error_msg = f"æ‰‹ç›¸åˆ†æå¤±è´¥: {str(e)}"
        yield f"data: {json.dumps({'type': 'error', 'error': error_msg}, ensure_ascii=False)}\n\n"


@router.post("/fortune/hand/analyze/stream", summary="æ‰‹ç›¸åˆ†æï¼ˆæµå¼ï¼‰")
async def analyze_hand_stream(
    image: UploadFile = File(..., description="æ‰‹æŒç…§ç‰‡"),
    solar_date: Optional[str] = Form(None, description="é˜³å†æ—¥æœŸ YYYY-MM-DD"),
    solar_time: Optional[str] = Form(None, description="å‡ºç”Ÿæ—¶é—´ HH:MM"),
    gender: Optional[str] = Form(None, description="æ€§åˆ« male/female"),
    use_bazi: bool = Form(True, description="æ˜¯å¦ç»“åˆå…«å­—åˆ†æ"),
    api_key: str = Depends(verify_api_key)
):
    """
    æ‰‹ç›¸åˆ†ææ¥å£ï¼ˆæµå¼ç‰ˆæœ¬ï¼‰
    ä½¿ç”¨ Server-Sent Events (SSE) å®æ—¶è¿”å›åˆ†æè¿›åº¦å’Œç»“æœ
    
    éœ€è¦è®¤è¯ï¼šåœ¨è¯·æ±‚å¤´ä¸­æä¾› X-API-Key
    """
    try:
        # è¯»å–å›¾åƒæ•°æ®
        image_bytes = await image.read()
        image_format = image.filename.split('.')[-1] if image.filename else "jpg"
        
        return StreamingResponse(
            analyze_hand_stream_generator(
                image_bytes, image_format, solar_date, solar_time, gender, use_bazi
            ),
            media_type="text/event-stream",
            headers={
                "Cache-Control": "no-cache",
                "Connection": "keep-alive",
                "X-Accel-Buffering": "no"  # ç¦ç”¨ nginx ç¼“å†²
            }
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"æ‰‹ç›¸åˆ†æå¤±è´¥: {str(e)}")


async def analyze_face_stream_generator(image_bytes: bytes, image_format: str,
                                       solar_date: Optional[str], solar_time: Optional[str],
                                       gender: Optional[str], use_bazi: bool):
    """é¢ç›¸åˆ†ææµå¼ç”Ÿæˆå™¨ï¼ˆçœŸæ­£çš„æµå¼å“åº”ï¼‰"""
    try:
        # å‘é€å¼€å§‹æ¶ˆæ¯ï¼ˆåŒ…å«æ˜¾ç¤ºæŒ‡ä»¤ï¼‰
        start_msg = {
            'type': 'start',
            'statusText': 'å¼€å§‹åˆ†æé¢ç›¸',
            'showGrid': True,
            'showStatus': True,
            'showPlayButton': True
        }
        yield f"data: {json.dumps(start_msg, ensure_ascii=False)}\n\n"
        await asyncio.sleep(0)
        
        # ä½¿ç”¨é¢ç›¸æµå¼åˆ†æå™¨ï¼ˆçœŸæ­£çš„æµå¼å“åº”ï¼‰
        if FACE_STREAM_ANALYZER_AVAILABLE and face_analyzer_stream:
            logger.info("âœ… ä½¿ç”¨é¢ç›¸æµå¼åˆ†æå™¨ï¼ˆçœŸæ­£çš„æµå¼å“åº”ï¼‰")
            try:
                # ä½¿ç”¨é¢ç›¸æµå¼åˆ†æå™¨
                async for result in face_analyzer_stream.analyze_face_stream(
                    image_bytes=image_bytes,
                    image_format=image_format,
                    solar_date=solar_date,
                    solar_time=solar_time,
                    gender=gender,
                    use_bazi=use_bazi,
                    progress_callback=lambda text: None  # è¿›åº¦åœ¨å†…éƒ¨å¤„ç†
                ):
                    logger.debug(f"ğŸ“¤ æµå¼åˆ†æå™¨è¿”å›: type={result.get('type')}, statusText={result.get('statusText', '')}")
                    
                    # è½¬æ¢ç»“æœæ ¼å¼
                    if result.get('type') == 'progress':
                        progress_msg = {
                            'type': 'progress',
                            'statusText': result.get('statusText', 'æ­£åœ¨åˆ†æ...'),
                            'showGrid': True,
                            'showStatus': True,
                            'showPlayButton': True
                        }
                        yield f"data: {json.dumps(progress_msg, ensure_ascii=False)}\n\n"
                        await asyncio.sleep(0)
                    elif result.get('type') == 'complete':
                        data = result.get('data', {})
                        logger.debug(f"ğŸ“¤ å‘é€completeæ¶ˆæ¯ï¼Œdataç±»å‹: {type(data)}, success: {data.get('success') if isinstance(data, dict) else 'N/A'}")
                        complete_msg = {
                            'type': 'complete',
                            'data': data,
                            'statusText': 'åˆ†æå®Œæˆ',
                            'showGrid': False,
                            'showStatus': False,
                            'showPlayButton': False
                        }
                        yield f"data: {json.dumps(complete_msg, ensure_ascii=False)}\n\n"
                        logger.debug("âœ… completeæ¶ˆæ¯å·²å‘é€åˆ°å®¢æˆ·ç«¯")
                    elif result.get('type') == 'error':
                        error_msg = {
                            'type': 'error',
                            'error': result.get('error', 'åˆ†æå¤±è´¥'),
                            'statusText': 'åˆ†æå¤±è´¥',
                            'showGrid': False,
                            'showStatus': False,
                            'showPlayButton': False
                        }
                        yield f"data: {json.dumps(error_msg, ensure_ascii=False)}\n\n"
                        logger.error("âŒ erroræ¶ˆæ¯å·²å‘é€åˆ°å®¢æˆ·ç«¯")
            except Exception as stream_error:
                import traceback
                error_detail = f"æµå¼åˆ†æå™¨å¼‚å¸¸: {str(stream_error)}\n{traceback.format_exc()}"
                logger.error(f"âŒ {error_detail}")
                error_msg = {
                    'type': 'error',
                    'error': f"æµå¼åˆ†æå¤±è´¥: {str(stream_error)}",
                    'statusText': 'åˆ†æå¤±è´¥',
                    'showGrid': False,
                    'showStatus': False,
                    'showPlayButton': False
                }
                yield f"data: {json.dumps(error_msg, ensure_ascii=False)}\n\n"
        else:
            logger.warning("âš ï¸  æµå¼åˆ†æå™¨ä¸å¯ç”¨ï¼Œä½¿ç”¨ gRPCï¼ˆéæµå¼ï¼Œè¾ƒæ…¢ï¼‰")
            # é™çº§æ–¹æ¡ˆï¼šä½¿ç”¨ gRPCï¼ˆéæµå¼ï¼‰
            # å…ˆéªŒè¯å›¾åƒ
            if IMAGE_VALIDATOR_AVAILABLE and image_validator:
                logger.debug("ğŸ” éªŒè¯å›¾åƒ...")
                is_valid, error_msg = image_validator.validate_face_image(image_bytes)
                logger.debug(f"éªŒè¯ç»“æœ: æœ‰æ•ˆ={is_valid}, é”™è¯¯={error_msg}")
                if not is_valid:
                    error_response = {
                        'type': 'error',
                        'error': error_msg,
                        'statusText': 'å›¾åƒéªŒè¯å¤±è´¥',
                        'showGrid': False,
                        'showStatus': False,
                        'showPlayButton': False
                    }
                    yield f"data: {json.dumps(error_response, ensure_ascii=False)}\n\n"
                    return
            else:
                logger.warning("âš ï¸  å›¾åƒéªŒè¯å™¨ä¸å¯ç”¨ï¼Œè·³è¿‡éªŒè¯")
            
            # æ„å»º gRPC è¯·æ±‚
            request = fortune_analysis_pb2.FaceAnalysisRequest()
            request.image.image_bytes = image_bytes
            request.image.image_format = image_format
            request.image.filename = f"face.{image_format}"
            
            # å…«å­—ä¿¡æ¯
            if use_bazi and solar_date and solar_time and gender:
                request.bazi_info.solar_date = solar_date
                request.bazi_info.solar_time = solar_time
                request.bazi_info.gender = gender
                request.bazi_info.use_bazi = True
            else:
                request.bazi_info.use_bazi = False
            
            # è°ƒç”¨ gRPC æœåŠ¡
            address = get_grpc_client()
            
            with grpc.insecure_channel(address) as channel:
                stub = fortune_analysis_pb2_grpc.FortuneAnalysisServiceStub(channel)
                response = stub.AnalyzeFace(request, timeout=60.0)
            
            if response.success:
                report = json.loads(response.report_json) if response.report_json else {}
                complete_msg = {
                    'type': 'complete',
                    'data': report,
                    'statusText': 'åˆ†æå®Œæˆ',
                    'showGrid': False,
                    'showStatus': False,
                    'showPlayButton': False
                }
                yield f"data: {json.dumps(complete_msg, ensure_ascii=False)}\n\n"
            else:
                error_msg = {
                    'type': 'error',
                    'error': response.error_message,
                    'statusText': 'åˆ†æå¤±è´¥',
                    'showGrid': False,
                    'showStatus': False,
                    'showPlayButton': False
                }
                yield f"data: {json.dumps(error_msg, ensure_ascii=False)}\n\n"
            
    except Exception as e:
        import traceback
        error_msg = f"é¢ç›¸åˆ†æå¤±è´¥: {str(e)}"
        yield f"data: {json.dumps({'type': 'error', 'error': error_msg}, ensure_ascii=False)}\n\n"


@router.post("/fortune/face/analyze/stream", summary="é¢ç›¸åˆ†æï¼ˆæµå¼ï¼‰")
async def analyze_face_stream(
    image: UploadFile = File(..., description="å¤´éƒ¨é¢ç›¸ç…§ç‰‡"),
    solar_date: Optional[str] = Form(None, description="é˜³å†æ—¥æœŸ YYYY-MM-DD"),
    solar_time: Optional[str] = Form(None, description="å‡ºç”Ÿæ—¶é—´ HH:MM"),
    gender: Optional[str] = Form(None, description="æ€§åˆ« male/female"),
    use_bazi: bool = Form(True, description="æ˜¯å¦ç»“åˆå…«å­—åˆ†æ"),
    api_key: str = Depends(verify_api_key)
):
    """
    é¢ç›¸åˆ†ææ¥å£ï¼ˆæµå¼ç‰ˆæœ¬ï¼‰
    ä½¿ç”¨ Server-Sent Events (SSE) å®æ—¶è¿”å›åˆ†æè¿›åº¦å’Œç»“æœ
    
    éœ€è¦è®¤è¯ï¼šåœ¨è¯·æ±‚å¤´ä¸­æä¾› X-API-Key
    """
    try:
        image_bytes = await image.read()
        image_format = image.filename.split('.')[-1] if image.filename else "jpg"
        
        return StreamingResponse(
            analyze_face_stream_generator(
                image_bytes, image_format, solar_date, solar_time, gender, use_bazi
            ),
            media_type="text/event-stream",
            headers={
                "Cache-Control": "no-cache",
                "Connection": "keep-alive",
                "X-Accel-Buffering": "no"
            }
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"é¢ç›¸åˆ†æå¤±è´¥: {str(e)}")

